File: go/middleware.go
package l402kit

import (
	"context"
	"encoding/json"
	"fmt"
	"net/http"
)

// Invoice represents a Lightning payment request created by a provider.
type Invoice struct {
	PaymentRequest string
	PaymentHash    string
	Macaroon       string
	AmountSats     int
}

// LightningProvider is the interface for creating Lightning invoices.
// Implement this to plug in any Lightning backend.
type LightningProvider interface {
	CreateInvoice(ctx context.Context, amountSats int) (Invoice, error)
}

// Options configures the L402 middleware.
type Options struct {
	// PriceSats is the cost per API call in satoshis.
	PriceSats int

	// Lightning is your Lightning provider — required.
	// Use NewBlinkProvider, NewAlbyProvider, NewBtcPayProvider, or NewManagedProvider.
	Lightning LightningProvider

	// OnPayment is an optional callback called after a valid payment is verified.
	OnPayment func(token L402Token, amountSats int)

	// Deprecated: use Lightning: NewManagedProvider(address) instead.
	OwnerLightningAddress string
}

// Middleware returns an http.Handler that enforces L402 payment before calling next.
//
//	blink := l402kit.NewBlinkProvider(os.Getenv("BLINK_API_KEY"), os.Getenv("BLINK_WALLET_ID"))
//	mux.Handle("/api/data", l402kit.Middleware(l402kit.Options{
//	    PriceSats: 10,
//	    Lightning: blink,
//	}, myHandler))
func Middleware(opts Options, next http.Handler) http.Handler {
	if opts.Lightning == nil {
		panic(ErrNoProvider)
	}

	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		auth := r.Header.Get("Authorization")

		if len(auth) > 5 && auth[:5] == "L402 " {
			token := auth[5:]
			valid, _ := VerifyToken(token)

			if valid {
				t, _ := ParseToken(token)

				if !CheckAndMarkPreimage(t.Preimage) {
					writeJSON(w, http.StatusUnauthorized, map[string]string{"error": "Token already used"})
					return
				}

				if opts.OnPayment != nil {
					opts.OnPayment(t, opts.PriceSats)
				}

				next.ServeHTTP(w, r)
				return
			}
		}

		// No valid token — create invoice and return 402
		inv, err := opts.Lightning.CreateInvoice(r.Context(), opts.PriceSats)
		if err != nil {
			writeJSON(w, http.StatusInternalServerError, map[string]string{"error": err.Error()})
			return
		}

		w.Header().Set("WWW-Authenticate", fmt.Sprintf(
			`L402 macaroon="%s", invoice="%s"`,
			inv.Macaroon, inv.PaymentRequest,
		))
		writeJSON(w, http.StatusPaymentRequired, map[string]any{
			"error":     "Payment Required",
			"priceSats": opts.PriceSats,
			"invoice":   inv.PaymentRequest,
			"macaroon":  inv.Macaroon,
		})
	})
}

func writeJSON(w http.ResponseWriter, status int, v any) {
	w.Header().Set("Content-Type", "application/json")
	w.WriteHeader(status)
	json.NewEncoder(w).Encode(v) //nolint:errcheck
}

File: rust/src/middleware.rs
use axum::{
    extract::{Request, State},
    http::{header, StatusCode},
    middleware::Next,
    response::{IntoResponse, Response},
    Json,
};
use serde_json::json;
use std::sync::Arc;

use crate::{
    replay::check_and_mark_preimage,
    types::Options,
    verify::{parse_token, verify_token},
};

/// Axum middleware that enforces L402 payment before calling the next handler.
///
/// ```rust,no_run
/// use axum::{Router, routing::get, middleware};
/// use l402kit::{l402_middleware, ManagedProvider, Options};
/// use std::sync::Arc;
///
/// async fn my_handler() -> &'static str { "ok" }
///
/// let provider = ManagedProvider::new("you@yourdomain.com".into());
/// let opts = Arc::new(Options::new(10, provider));
/// let app: Router<()> = Router::new()
///     .route("/api/data", get(my_handler))
///     .route_layer(middleware::from_fn_with_state(opts, l402_middleware));
/// ```
pub async fn l402_middleware(
    State(opts): State<Arc<Options>>,
    request: Request,
    next: Next,
) -> Response {
    let auth = request
        .headers()
        .get(header::AUTHORIZATION)
        .and_then(|v| v.to_str().ok())
        .map(str::to_string);

    if let Some(auth) = auth {
        if let Some(token) = auth.strip_prefix("L402 ") {
            if verify_token(token) {
                if let Ok(t) = parse_token(token) {
                    if !check_and_mark_preimage(&t.preimage) {
                        return (
                            StatusCode::UNAUTHORIZED,
                            Json(json!({ "error": "Token already used" })),
                        )
                            .into_response();
                    }

                    if let Some(cb) = &opts.on_payment {
                        cb(t, opts.price_sats);
                    }

                    return next.run(request).await;
                }
            }
        }
    }

    // No valid token — create invoice and return 402
    match opts.lightning.create_invoice(opts.price_sats).await {
        Ok(inv) => {
            let www_auth = format!(
                r#"L402 macaroon="{}", invoice="{}""#,
                inv.macaroon, inv.payment_request
            );
            let mut response = (
                StatusCode::PAYMENT_REQUIRED,
                Json(json!({
                    "error": "Payment Required",
                    "priceSats": opts.price_sats,
                    "invoice": inv.payment_request,
                    "macaroon": inv.macaroon,
                })),
            )
                .into_response();
            response.headers_mut().insert(
                header::WWW_AUTHENTICATE,
                www_auth.parse().unwrap(),
            );
            response
        }
        Err(e) => (
            StatusCode::INTERNAL_SERVER_ERROR,
            Json(json!({ "error": e.to_string() })),
        )
            .into_response(),
    }
}
